### ISSUE TITLE
floor_divide on float16/bfloat16 depends on tensor length, and the vectorized result can exceed the quotient

### ISSUE THREAD

**Issue by krrishapatel** [pytorch/pytorch#193754]
### 🐛 Describe the bug

`torch.floor_divide` on float16 and bfloat16 gives a different answer depending on how long the tensor is, and the vectorized answer can be larger than the quotient it is supposed to be the floor of.

```python
import torch

a = torch.full((2,), 560.0, dtype=torch.bfloat16)
b = torch.full((2,), 3.0, dtype=torch.bfloat16)
print(torch.floor_divide(a, b)[0].item())        # 187.0
print(torch.floor_divide(a[:1], b[:1])[0].item())  # 186.0
```

560 and 3 are both exact in bfloat16, and 560 / 3 is 186.67, so the floor is 186. The length-2 result 187 gives `187 * 3 = 561`, which is greater than 560. The reconstruction breaks with it:

```python
q = torch.floor_divide(a, b)
r = torch.remainder(a, b)
print((q * b + r)[0].item())  # 563.0, not 560.0
```

float16 does the same:

```python
a = torch.full((2,), 2158.0, dtype=torch.float16)
b = torch.full((2,), 3.0, dtype=torch.float16)
print(torch.floor_divide(a, b)[0].item())        # 718.0
print(torch.floor_divide(a[:1], b[:1])[0].item())  # 719.0   correct
```

On 4000 random pairs per dtype, comparing a length-1 tensor against a length-64 tensor of the same value:

| dtype | pairs that disagree |
|---|---|
| float16 | 39 / 4000 |
| bfloat16 | 65 / 4000 |
| float32 | 0 / 4000 |
| float64 | 0 / 4000 |

float32 and float64 being clean is what points at the cause. The scalar path calls `c10::div_floor_floating(a, b)` with `scalar_t` = `Half` or `BFloat16`, and `std::fmod` on those promotes to float, so the whole computation runs in fp32 and only the return value narrows:

https://github.com/pytorch/pytorch/blob/04b971abedf3167a3f929e4e7a9a5b7e8647c02c/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp#L341

The vectorized path runs in `Vectorized<Half>` / `Vectorized<BFloat16>`, so `a.fmod(b)` and `(a - mod) / b` each round back to 8 or 11 mantissa bits:

https://github.com/pytorch/pytorch/blob/04b971abedf3167a3f929e4e7a9a5b7e8647c02c/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp#L344

This class is already recognized in the same file. The SVE bf16 specialization does exactly the needed widening, for the reason given in its comment:

https://github.com/pytorch/pytorch/blob/04b971abedf3167a3f929e4e7a9a5b7e8647c02c/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp#L277-L295

It is guarded by `CPU_CAPABILITY_SVE256 && __ARM_FEATURE_BF16`, so every other build gets the narrow computation, and float16 gets it everywhere. Making that specialization unconditional and adding one for `Half` would line the vec path up with the scalar path.

I know results are not guaranteed bitwise identical across vectorization. This is not a last-bit difference. The result is off by a whole unit and is not a floor of anything.

Found by a differential tester that varies tensor length while holding the values fixed. Related to #193753, which is a different root cause in the same file.

### Versions

torch 2.13.0, CPU, macOS arm64.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01

**Comment by krrishapatel** [comment]
PR: #193766.

Correcting my own diagnosis above. I said the scalar path promotes to float because `std::fmod` does. It does not. `c10::BFloat16` has its own `std::fmod` overload that returns `c10::BFloat16`, so `c10::div_floor_floating` rounds at every step, and inside the tensor-tensor branch the scalar loop and the vectorized loop actually agree exactly. I measured 0 disagreements between them over 32000 pairs.

The real split is the other branch of `div_floor_kernel`. When the divisor is a scalar it dispatches through `AT_DISPATCH_REDUCED_FLOATING_TYPES` and computes in `opmath_t`. A size-1 tensor is squeezed to 0-dim, `iter.is_scalar(2)` becomes true, and it takes that path. That is why length 1 differs from length 2.

So widening only the vectorized loop, which is what I suggested, would have moved the inconsistency instead of fixing it. The PR widens both loops, plus the same one line on CUDA, which has the same split against its `is_cpu_scalar(2)` branch.

Two things I did not know when I filed this. The error is not bounded to one unit, the largest gap from float math over 4096 random values was 64. And the fix is about 9x faster for both dtypes, because `Vectorized<BFloat16>::fmod` is `map2(q, std::fmod)` and `::floor` is `map(floor_impl)`, so the old path called scalar libm once per lane.

**Comment by yun520-1** [comment]
Thanks for the detailed vectorized-path report. The length-dependent result on `floor_divide` for bf16/fp16 is a very concrete symptom: it points to the vectorized kernel accumulating rounding error differently than the scalar fallback, which makes the quotient exceed the true floor only when the intermediate product stays in low precision across more elements.

A discriminator-style observation: this is exactly the class of bug that is invisible until someone compares `a[:1]` vs `a`. Most tests use scalar or small-tensor paths, so the vectorized fp16/bf16 path silently drifts. Two quick mitigations while a proper fix lands:
1. Accumulate the division in float32 inside the kernel and cast the floor result back to the original dtype.
2. Add a tolerance-based correctness check for fp16/bf16 arithmetic ops in the CI vectorized test matrix.

If the root cause is `std::fmod` behavior on bf16, prefer explicit `floor(x / y)` via float32 promotion rather than relying on the current vectorized remainder path.

HeartFlow discriminators would flag this as a **numerical-stability** pattern — vectorized low-precision ops that depend on tensor length are a recurring failure mode in ML frameworks, not a one-off PyTorch bug.

**Comment by krrishapatel** [comment]
The fix is up as #193766. It does the float32 widening you describe, in both the vectorized loop and the scalar loop.

Two corrections though.

There is no accumulation across elements. This is elementwise, and inside the tensor-tensor branch the vectorized loop and the scalar loop agree exactly. I measured 0 disagreements over 32000 pairs. The split is a different branch of `div_floor_kernel`, which already computes in `opmath_t` when the divisor is a scalar. A size-1 tensor gets squeezed to 0-dim and reaches that branch, which is the whole reason `a[:1]` differs from `a`. See my comment above.

`floor(x / y)` is not equivalent and would trade this for a different bug. See NOTE: [Floor Division in Python] in `c10/util/generic_math.h`. The quotient is computed as `(a - fmod(a, b)) / b` on purpose, to keep `a == (a // b) * b + remainder(a, b)`, which plain `floor(a / b)` does not.

On the test, tolerance is the wrong tool here. bfloat16 default `rtol` is about 1.6e-2, and an off-by-one quotient at 186 is 0.5%, so a tolerance check passes while the bug is present. The test in the PR compares against float math narrowed once with `atol=0, rtol=0`, and covers lengths 1, 2 and 3 so both loops are exercised.
